File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        holder=None
        while root:
            if root.right:
                tempHolder=holder
                holder=root.right
                temp=holder
                while temp.right:
                    temp=temp.right
                temp.right=tempHolder
            if root.left:
                root.right=root.left
                root.left=None
            else:
                root.right=holder
                holder=None
            root=root.right
        

File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    
    TreeNode *dfs(TreeNode *root){
        if(!root->left && !root->right) return root;
        if(root->left) swap(root->left,root->right);
        TreeNode *lastPoint=dfs(root->right);
        if(root->left){
            lastPoint->right=root->left;
            root->left=NULL;
            lastPoint=dfs(lastPoint->right);
        }
        return lastPoint;
    }
    
public:
    void flatten(TreeNode* root) {
        if(!root) return;
        dfs(root);
    }
};